blog/2025-08-22-pattern.mdx

```diff
@@ -0,0 +1,312 @@
+named identifier after them. For example, `[x, y, ,]` has length 2 but
+`[x, , y]` has length 3.
+
+:::note
+If you want to enforce length 'l' of a tuple pattern you must have a named identifier
+at the `l-1` index of your pattern. For example, if you need to enforce a tuple
+pattern of legth 3 you need to have a named identifier on index 2 (e.g., all
+`[, , x]`, `[, x, y]`, and `[x, y, z]` enforce a tuple pattern of length 3).
+:::
+
+### Type Annotations for Tuple Patterns
+
+In a circuit definition, type declarations are required for tuple patterns:
+
+```compact
+circuit exampleTuplePattern([x, y]: [Field, Boolean]): Field {
+  // x has type Field, y has type Boolean
+  return x;
+}
+```
+
+Such a type declaration dictates the structure and the type of each element.  For example,
+in `exampleTuplePattern` the input has to be a tuple, it first element has to be a value
+of type `Field`, and its second element has to be a value of type `Boolean`.
+A call to `exampleTuplePattern` that does not meet these restrictions
+(e.g., `exampleTuplePattern([1, 1])`) will trigger a type error.
+
+For const bindings and anonymous circuits, type annotations are optional but recommended
+for clarity and an easier debugging experience. When a type annotation is provided
+ensure its structure aligns with the pattern. For example, `[a, , b] : [Field, Field, Field]`
+aligns the structure of the pattern and its type whereas `[a, , b] : [Field, Field]` does not:
+the pattern states that it is a tuple with three elements but the type annotation is
+tuple type of length two.
+
+Unlike patterns, types do not allow skipping elements in a tuple type. For example,
+`[a, , b] : [Field, , Field]` causes a parse error. So even when you're skipping an
+element in a pattern you must assign it a dummy type. Unfortunately, you have to
+remember what dummy type you used when you're passing arguments to pattern parameters.
+For example,
+
+```compact
+circuit haveToUseDummyType([, x]: [Field, Boolean]): Boolean {
+  return x;
+}
+
+export circuit callSite(): Boolean {
+  return haveToUseDummyType([1, true]);
+
+  // return haveToUseDummyType([true, true]);
+  // triggers a type error since the first element must be
+  // a value of type Field
+}
+```
+
+:::note
+We recommend you using a sepecific type as your dummy type throughout all your Compact
+code or use a type synonym for such a dummy type when such a feature is available. For
+example, always use `Boolean` as your dummy type.
+:::
+
+Similar to tuple types, skipped elements of a tuple value cannot be dropped, for example,
+`const [x, , y] : [Field, Field, Field] = [1, , 3];` triggers a type error.
+
+When a type annotation exists, the length of the tuple pattern must be less than or equal to
+the length of its tuple type annotation.
+For example, `[a, b]: [Field, Field]` typechecks.
+So does `[, a]: [Field, Field]`. So does `[a, , , , ,] : [Field, Field]`.
```

Comment:
I *think* this is best explained as: we allow a trailing comma in tuples because that can be convenient.  We treat any number of trailing commas the same as a single one.  So the pattern `[a, , , , ,]` is equivalent to `[a,]` which is equivalent to `[a]`.

Reply:
And then, after removing the trailing commas, you introduce fresh identifiers for all the missing elements.  So `[, , a]` is equivalent to `[, , a]` (itself) after removing all the (zero) trailing commas.  And that's equivalent to `[_0, _1, a]` provided those names are fresh.